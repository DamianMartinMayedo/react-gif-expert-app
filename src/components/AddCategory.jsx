import { useState } from "react"


export const AddCategory = ({ onNewCategory }) => {

    const [inputValue, setinputValue] = useState('')

    //funcion para detectar un cambio en el input
    const onInputChange = (event) => {
        //asi obtenemos lo que el usuario está escribiendo
        setinputValue(event.target.value);
        // console.log(event.target.value);
    }

    //
    //prevenimos el evento del form y enviamos los datos al componente padre para añadir a la lista
    const onSubmit = (event) => {
        event.preventDefault();
        //evitar que el input este vacio o sea una sola letra
        if (inputValue.trim().length <= 1) return;
        //añadir a la lista de la primera manera
        // setCategories(categories => [inputValue, ...categories]);
        
        //se añade a la lista de manera más eficiente, la funcion manejadora está en el componente padre
        onNewCategory(inputValue.trim());
        //limpiar el input luego del enter
        setinputValue('');
    }

    return (
        <form onSubmit={onSubmit}>
            <input
                type="text"
                placeholder="buscar Gifs"
                value={inputValue}
                onChange={onInputChange}
            />
        </form>
    )
}
