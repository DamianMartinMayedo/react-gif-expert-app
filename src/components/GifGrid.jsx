import PropTypes from 'prop-types';
import { GifItem } from "./GifItem";
import { useFetchGifs } from "../hooks/useFetchGifs";


export const GifGrid = ({ category }) => {

    //esto es un custom hook (creado por mi) se utiliza "use" de manera de guia
    //permite reutilizar codigo de manera muy sencilla

    const { images, isLoading } = useFetchGifs(category);

    //----------------------------------------------------------------------

    return (
        <>
            {
                isLoading && (<h2>Cargando......</h2>)
            }
            <h3>{category}</h3>


            {/* En react se utiliza className en vez de class (reservada en JS para crear clases) */}
            <div className="card-grid">
                {
                    images.map((img) => (
                        <GifItem
                            key={img.id}
                            // utilizando el spread se "esparsen" todas las propiedades de image hacia el componente
                            // el componente las recibe toda y luego en sus props decidimos cuales utilizar
                            // se utiliza fundamentalmenyÂ¡te cuando couando hay muchas propiedes
                            {...img}
                        />
                    ))
                }

            </div>

        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
}