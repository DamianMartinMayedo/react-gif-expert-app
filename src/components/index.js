//este archivo permite agrupar las exportaciones de archivos en la misma ruta
//se le llama archivo de barril
//por defecto cuando se apunta a la carpeta primero se busca en el index 

export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';
