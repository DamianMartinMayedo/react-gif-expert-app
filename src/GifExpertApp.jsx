import { useState } from "react";
import { AddCategory, GifGrid } from "./components";



export const GifExpertApp = () => {
    
    const [categories, setCategories] = useState(['AC Milan']);

    const onAddCategory = (newCategory) => {
        if (categories.includes(newCategory)) return;
        setCategories([newCategory, ...categories]);
    }

    return (
        <>

            <h1>GifExpertApp</h1>

            <AddCategory
                //enviando la funcion de aÃ±adir categoria  como prop
                onNewCategory={onAddCategory}
            />

            {
                categories.map((category) => (
                    <GifGrid
                        key={category}
                        category={category}
                    />
                ))
            }


        </>
    )
}
