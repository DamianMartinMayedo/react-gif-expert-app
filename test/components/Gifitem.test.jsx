import { render,screen } from "@testing-library/react"
import { GifItem } from "../../src/components"

describe('pruebas en Gifitem', () => { 
    const title = 'AC Milan';
    const url = 'https://loco.com/'
    test('snapshot', () => { 
        const {container}=render(<GifItem title={title} url={url}/>);
        expect(container).toMatchSnapshot();
    
    });
    test('debe mostrar url y alt indicado', () => { 
        render(<GifItem title={title} url={url}/>);
        // expect(screen.getByRole('img').src).toBe(url);
        const{src, alt} = screen.getByRole('img');
        expect(src).toBe(url);
        expect(alt).toBe(title);

     });

     test('existe el titulo?', () => { 
        render(<GifItem title={title} url={url}/>);
        expect(screen.getByText(title)).toBeTruthy();
     })
});