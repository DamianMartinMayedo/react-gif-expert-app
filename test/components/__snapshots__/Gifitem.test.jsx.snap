// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en Gifitem snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="AC Milan"
      src="https://loco.com/"
    />
    <p>
      AC Milan
    </p>
  </div>
</div>
`;
